apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RemoveMe/MediaDownloader.cs

```diff
@@ -0,0 +1,355 @@
+        /// </summary>
+        public int ChunkSize
+        {
+            get { return chunkSize; }
+            set
+            {
+                if (value > MaximumChunkSize)
+                {
+                    throw new ArgumentOutOfRangeException("ChunkSize");
+                }
+                chunkSize = value;
+            }
+        }
+
+        /// <summary>
+        /// The range header for the request, if any. This can be used to download specific parts
+        /// of the requested media.
+        /// </summary>
+        public RangeHeaderValue Range { get; set; }
+
+        #region Progress
+
+        /// <summary>
+        /// Download progress model, which contains the status of the download, the amount of bytes whose where 
+        /// downloaded so far, and an exception in case an error had occurred.
+        /// </summary>
+        private class DownloadProgress : IDownloadProgress
+        {
+            /// <summary>Constructs a new progress instance.</summary>
+            /// <param name="status">The status of the download.</param>
+            /// <param name="bytes">The number of bytes received so far.</param>
+            public DownloadProgress(DownloadStatus status, long bytes)
+            {
+                Status = status;
+                BytesDownloaded = bytes;
+            }
+
+            /// <summary>Constructs a new progress instance.</summary>
+            /// <param name="exception">An exception which occurred during the download.</param>
+            /// <param name="bytes">The number of bytes received before the exception occurred.</param>
+            public DownloadProgress(Exception exception, long bytes)
+            {
+                Status = DownloadStatus.Failed;
+                BytesDownloaded = bytes;
+                Exception = exception;
+            }
+
+            /// <summary>Gets or sets the status of the download.</summary>
+            public DownloadStatus Status { get; private set; }
+
+            /// <summary>Gets or sets the amount of bytes that have been downloaded so far.</summary>
+            public long BytesDownloaded { get; private set; }
+
+            /// <summary>Gets or sets the exception which occurred during the download or <c>null</c>.</summary>
+            public Exception Exception { get; private set; }
+        }
+
+        /// <summary>
+        /// Updates the current progress and call the <see cref="ProgressChanged"/> event to notify listeners.
+        /// </summary>
+        private void UpdateProgress(IDownloadProgress progress)
+        {
+            ProgressChanged?.Invoke(progress);
+        }
+
+        #endregion
+
+        /// <summary>Constructs a new downloader with the given client service.</summary>
+        public TmpMediaDownloader(IClientService service)
+        {
+            this.service = service;
+        }
+
+        #region IMediaDownloader Overrides
+
+        public event Action<IDownloadProgress> ProgressChanged;
```

Comment:
I've never seen this pattern before. Is this ok to be public? Should we worry about the possibility that there may be more information to send in the future, which might make a derived EventArgs a better approach, or would that just be added to IDownloadProgress?